Add `--bootstrap`

cknoll opened the pull request:
- move the directories `examples` and `templates` inside `src/exam_generator` and let them install inside `sitepackages` during `pip install .`
- add command `--bootstrap`: copy `examples` and `templates` to local working directory

cknoll committed:
add metavar args to the parser.add_argument-statements

This makes the exam_generator --help easier to understand

cknoll committed:
add bootstrapping functionality

cknoll committed:
moving templates, examples inside package dir

cknoll committed:
setup.py: add templates, examples as package_data

cknoll committed:
add basic boostrap command (dummy only yet)

jcwinkler commented:
I cannot observe the intended behaviour.

**Steps to reproduce**
- Clone repo
- Change into repo root
- Enter `pip install .`

**Observation**
- Python files only are installed under `Lib/site-packages/exam_generator`
- Furthermore, the following warning is written to the consolse
  ```bash
  Installing collected packages: exam-generator
  DEPRECATION: exam-generator is being installed using the legacy 'setup.py install' method,
  because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1
  will enforce this behaviour change. A possible replacement is to enable the '--use-pep517'
  option. Discussion can be found at https://github.com/pypa/pip/issues/8559
  ```

**System**
- Python 3.11.0
- Windows 10 21H2

**bootstrap option passed to exam_generator**
- ~~Where does this option needs to be passed?~~
- Fails due to error described abvoe

jcwinkler commented on setup.py:
```diff
@@ -18,22 +18,36 @@
+            paths.append(os.path.join(path, filename))
+    return paths
+
+package_file_list1 = package_files("examples")
+package_file_list2 = package_files("templates")
```
Here we are in the directory in which setup.py resides. Hence, nothing will be found. (Did you forget to delete the old dirs locally?)

jcwinkler commented on setup.py:
```diff
@@ -18,22 +18,36 @@
     requirements = requirements_file.read()
 
+
+def package_files(directory):
+    # source: https://stackoverflow.com/a/36693250
+    paths = []
+    for (path, directories, filenames) in os.walk(directory):
+        for filename in filenames:
+            paths.append(os.path.join(path, filename))
+    return paths
```
This function should return the list relative to the package directory, i,.e., relative to `src/exam_generator`.

jcwinkler commented on setup.py:
```diff
@@ -18,22 +18,36 @@
     packages=find_packages("src"),
     package_dir = {"": "src"},
-    package_data={'abc': ['templates/*']},
+    package_data={"": [*package_file_list1, *package_file_list2]},
```
Must be relative to `src/exam_generator`, see comments above.

jcwinkler reviewed (commented):
I suggest the following replacement of `package_files`:

```python
def package_files(package_dir, directory):
    paths = []
    for (path, directories, filenames) in os.walk(os.path.join(package_dir, packagename, directory)):
        for filename in filenames:
            paths.append(os.path.join(path.replace(package_dir + os.sep, ""), filename))
    return paths
```
(or an even more improved variant)

Then call it as

```python
package_file_list1 = package_files("src", "examples")
package_file_list2 = package_files("src", "templates")
```

This will fix the issues mentioned below.

cknoll committed:
delete symbolic links

they are more confusing than helpful
they where included to make this content easiser browsable in github but this worked poorly anyway

cknoll committed:
setup.py: fix path problem

insert the suggested change from https://github.com/TUD-RST/examgenerator/pull/13#pullrequestreview-1180767268

cknoll committed:
add test for bootstrapping to ci

jcwinkler committed:
Merge pull request #15 from TUD-RST/feature/test_bootstrap_ci

Add test for bootstrapping to ci

cknoll committed:
add interactive debugging feature to ci (python-app-yml)

cknoll committed:
debug ci

cknoll committed:
debug ci (2)

cknoll committed:
debug ci (3)

cknoll committed:
debug ci (4)

cknoll committed:
debug ci (5)

cknoll committed:
remove ci debugging code

jcwinkler committed:
Merge pull request #16 from TUD-RST/feature/test_bootstrap_ci

Feature/test bootstrap ci (2nd try)
